### DISCUSSION TITLE
C++ suggestions for btop

### DISCUSSION THREAD

**Discussion by WSLUser** [aristocratos/bpytop#338] [Ideas]
Your README states provide some suggestions for C++ and I have a few libraries I'd suggest you check out: 

1st: I'd suggest making use of CLI11 for your argument parser: https://github.com/CLIUtils/CLI11.

2nd: Secondly, if you need to use any math (which I think would likely be needed to retrieve certain information), then use the safe math library Chromium math: https://github.com/chromium/chromium/tree/master/base/numerics

3rd: Make use of https://github.com/pinam45/dynamic_bitset and it's optional dependency: https://github.com/kimwalisch/libpopcnt

4th: Make use of https://github.com/fmtlib/fmt. You'll be passing alot of strings around so this library makes alot of sense to use from a performance and safety perspective as well as offering similar syntax to python for string formatting though you might be able to get other benefits from this library as well.

Lastly:  Make use of https://github.com/microsoft/GSL for the parts that make sense to use.

**Comment by aristocratos** [comment]
Thanks for the suggestions but the linux version of btop is about 80% complete.
Cpu, memory, disks and processes collection and drawing is done, will publish the beta when network collect and draw is done, probably this weekend or the next.
The stuff missing for a complete release for linux is mostly the menus and some small stuff.
Some new stuff have already been added, like 3 options for the graph symbols (braille, blocks, and tty blocks) which can be selected independently for each box, full support for 16-color tty terminals (with auto detection), all text input have full editing capability (moving the cursor around and deleting from the middle of the string, etc.), some changes to the UI design and more.

The only non STL includes currently are `unordered_flat_map` from https://github.com/martinus/robin-hood-hashing, to get some better performing maps.
And relying mostly on the standard library (using STL algorithms where possible) seems to have been a good choice, since it's really fast, like faster than `htop` fast. And I'm counting on it only getting faster with community contributed optimizations. (This is the first C++ I've ever written so there will definitively be mistakes that could possibly be slowing stuff down).

`dynamic_bitset` and `libpopcnt` looks interesting though, I'm currently trying to optimize out some unnecessary string parsing in loops using `std::bitset`.

A screenshot of btop++ in it's current state:
![Screenshot at 2021-08-12 20-37-32](https://user-images.githubusercontent.com/59659483/129250629-4a14a9df-b3d0-44c8-b945-1590420f1a2c.png)

**Comment by WSLUser** [comment]
I should note that FMT is part of the C++ 20 standard/implementation. It just is available for use in older C++ versions being provided as separate library and can't emphasize enough that it ought to be used by any C++ project. I mentioned string operations but it offers much more than that.

Also nice find for the hashing perf boost library. Being of MIT license will help decent bit too.

**Comment by aristocratos** [comment]
@WSLUser 
I ended up writing my own string formatting functions since I'm dealing with a lot of multi-byte UTF-8 characters. And needed to be able to distinguish between c-string length (number of bytes), utf-8 length (detection of multi-byte characters) and actual displayed length on screen (UTF-8 wide characters).

So I'm using regular `std::string` and my own string manipulation functions that can switch between treating any string as a regular string, utf-8 string or wide utf-8 string. And using `std::string_view` where possible when working on strings also seems to be helping.

**Comment by USLTD** [comment]
https://github.com/hamon-in/cpslib

In Psutil's GitHub repository it was mentioned that this is Psutil's to C porting attempt.

> **NOTE!**
Even though Library is written in C as I know C++ is C but with more features and is compatible with C meaning it can run any C code without issue on the other hand C can't run every C++ code.

**Reply by USLTD** [reply]
There is also alternative:
https://github.com/GrbavaCigla/psutil-cpp

**Reply by USLTD** [reply]
This is written in C++

**Comment by philip-h-dye** [comment]
In [decoupled C library, without python](https://github.com/giampaolo/psutil/issues/1247#) #1247, psutil's author recommended [sigar](https://github.com/hyperic/sigar).

**Reply by aristocratos** [reply]
@philip-h-dye 
[btop++](https://github.com/aristocratos/btop) has been out for 6 months and it's written from scratch without external dependencies.
So this discussion isn't that relevant anymore.

But thanks for the suggestion nonetheless :)
